reference-implementation/src/kem.rs

```diff
@@ -23,146 +24,133 @@ pub trait Kem {
-        use concrete_hybrid_kem::AsBytes;
-        let (ct, ss) = <K as concrete_hybrid_kem::EncapsDerand>::encaps_derand(pkR, randomness).unwrap();
-        (ss.as_bytes().to_vec(), ct.as_bytes().to_vec())
+    fn encap_derand(pkR: &Self::EncapsulationKey, randomness: &[u8]) -> (SharedSecret, Ciphertext) {
+        <K as EncapsDerand>::encaps_derand(pkR, randomness)
     }
 
-    fn decap(enc: &[u8], skR: &Self::DecapsulationKey) -> Vec<u8> {
-        use concrete_hybrid_kem::AsBytes;
-        let enc = K::Ciphertext::from(enc);
-        <K as concrete_hybrid_kem::Kem>::decaps(skR, &enc)
-            .unwrap()
-            .as_bytes()
-            .to_vec()
+    fn decap(enc: &Ciphertext, skR: &Self::DecapsulationKey) -> SharedSecret {
+        <K as concrete_hybrid_kem::kem::Kem>::decaps(skR, enc)
     }
 }
 
 pub struct MlKemWithId<K, const ID: u16>(core::marker::PhantomData<K>);
 
 impl<K, const ID: u16> Kem for MlKemWithId<K, ID>
 where
-    K: concrete_hybrid_kem::Kem + concrete_hybrid_kem::EncapsDerand,
+    K: concrete_hybrid_kem::kem::Kem + EncapsDerand,
 {
     const ID: [u8; 2] = ID.to_be_bytes();
-    const N_SECRET: usize = K::SHARED_SECRET_LENGTH;
-    const N_ENC: usize = K::CIPHERTEXT_LENGTH;
-    const N_PK: usize = K::ENCAPSULATION_KEY_LENGTH;
-    const N_SK: usize = K::SEED_LENGTH; // Use seed length for ML-KEM per spec
-    const N_SEED: usize = K::SEED_LENGTH;
-    const N_RANDOM: usize = K::RANDOMNESS_LENGTH;
+    const N_SECRET: usize = K::SHARED_SECRET_SIZE;
+    const N_ENC: usize = K::CIPHERTEXT_SIZE;
+    const N_PK: usize = K::ENCAPSULATION_KEY_SIZE;
+    const N_SK: usize = K::SEED_SIZE; // Use seed length for ML-KEM per spec
+    const N_SEED: usize = K::SEED_SIZE;
+    const N_RANDOM: usize = K::RANDOMNESS_SIZE;
 
-    type EncapsulationKey = <K as concrete_hybrid_kem::Kem>::EncapsulationKey;
-    type DecapsulationKey = <K as concrete_hybrid_kem::Kem>::DecapsulationKey;
+    type EncapsulationKey = EncapsulationKey;
+    type DecapsulationKey = DecapsulationKey;
 
     fn generate_key_pair(
         rng: &mut impl rand::CryptoRng,
     ) -> (Self::DecapsulationKey, Self::EncapsulationKey) {
-        let (ek, dk) = <K as concrete_hybrid_kem::Kem>::generate_key_pair(rng).unwrap();
+        let (dk, ek, _info) = <K as concrete_hybrid_kem::kem::Kem>::generate_key_pair(rng);
         (dk, ek)
     }
 
     fn derive_key_pair(ikm: &[u8]) -> (Self::DecapsulationKey, Self::EncapsulationKey) {
-        let (ek, dk) = <K as concrete_hybrid_kem::Kem>::derive_key_pair(ikm).unwrap();
+        let (dk, ek, _info) = <K as concrete_hybrid_kem::kem::Kem>::derive_key_pair(ikm);
         (dk, ek)
     }
 
     fn serialize_public_key(pkX: &Self::EncapsulationKey) -> Vec<u8> {
-        use concrete_hybrid_kem::AsBytes;
-        pkX.as_bytes().to_vec()
+        pkX.clone()
     }
 
     fn deserialize_public_key(pkXm: &[u8]) -> Self::EncapsulationKey {
```

Comment:
Given these constants, there's a scary option:


```suggestion
    fn deserialize_public_key(pkXm: &[u8; Self::N_PK]) -> Self::EncapsulationKey {
```

You'd have to redefine the trait to include the constants, of course.

That doesn't work with this using the Ciphertext alias from the ML-KEM crate though.

Reply:
Yeah, until we get nicer const generics, I'll settle for an `assert!()`.

Reply:
To be perfectly blunt, const generics are good enough for what you are seeking to do here.  They have been for a while.  You can't do sums or conditions, but for what you have, with constants for everything, you don't need that.  You could avoid the need for bounds-checking, if nothing else.